simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,427 @@
+        ]
+
+        assert len(beta_schedule) == 0, (
+            "Beta scheduling is handled by the `UpdateIRLS` directive."
+            "Remove the redundant `BetaSchedule` from your list of directives.",
+        )
+
+        spherical_scale = [isinstance(d, SphericalDomain) for d in directive_list]
+        if any(spherical_scale):
+            assert spherical_scale.index(True) < self_ind, (
+                "The directive 'SphericalDomain' must be before UpdateIRLS "
+                "in the directiveList"
+            )
+
+        return True
+
+    def stopping_criteria(self):
+        """
+        Check for stopping criteria of max_irls_iteration or minimum change.
+        """
+        phim_new = 0
+        for reg in self.reg.objfcts:
+            if isinstance(reg, (Sparse, BaseSparse)):
+                reg.model = self.invProb.model
+                phim_new += reg(reg.model)
+
+        # Check for maximum number of IRLS cycles
+        if self.metrics.irls_iteration_count == self.max_irls_iterations:
+            if self.verbose:
+                print(
+                    "Reach maximum number of IRLS cycles:"
+                    + " {0:d}".format(self.max_irls_iterations)
+                )
+            return True
+
+        # Check if the function has changed enough
+        f_change = np.abs(self.metrics.f_old - phim_new) / (self.metrics.f_old + 1e-12)
+
+        if (
+            f_change < self.f_min_change
+            and self.metrics.irls_iteration_count > 1
+            and np.abs(
+                1.0
+                - self.invProb.phi_d / self.misfit_from_chi_factor(self.chifact_target)
+            )
+            < self.misfit_tolerance
+        ):
+            print("Minimum decrease in regularization." + "End of IRLS")
+            return True
+
+        self.metrics.f_old = phim_new
+
+        return False
+
+
+class SphericalDomain(InversionDirective):
+    """
+    Directive to update the regularization weights to account for spherical
+    parameters in radian and SI.
+
+    The scaling applied to the regularization weights is based on the ratio
+    between the maximum value of the model and the maximum value of angles (pi).
+    """
+
+    def initialize(self):
+        self.update_scaling()
+
+    def endIter(self):
+        self.update_scaling()
+
+    def update_scaling(self):
+        """
+        Add an 'angle_scale' to the list of weights on the angle regularization for the
+        different block of models to account for units of radian and SI.
+        """
+        # TODO Need to establish a clearer connection between the regularizations
+        # and the model blocks. There is an assumption here that the first
+        # regularization controls the amplitude.
+        max_p = []
+        for reg in self.reg.objfcts[0].objfcts:
+            f_m = abs(reg.f_m(self.invProb.model))
+            max_p += [np.max(f_m)]
+
+        max_p = np.asarray(max_p).max()
```

Comment:
No need to cast into array.
```suggestion
        max_p = max(max_p)
```